projects/components/src/stepper/stepper.component.ts

```diff
@@ -104,26 +105,54 @@ export class StepperComponent implements AfterContentInit {
 
   public steps$?: Observable<StepperTabComponent[]>;
 
-  public ngAfterContentInit(): void {
-    this.steps$ = queryListAndChanges$(this.steps).pipe(map(list => list.toArray()));
-  }
+  private readonly isNextDisabledSubject: BehaviorSubject<boolean> = new BehaviorSubject<boolean>(false);
+  public isNextDisabled$: Observable<boolean> = this.isNextDisabledSubject.asObservable();
 
-  // TODO: can be made a directive
-  public isNextDisabled(stepper: MatStepper): boolean {
-    const selectedStep = stepper.steps.get(stepper.selectedIndex);
-    const formValid = selectedStep?.stepControl?.valid;
-    const stepCompleted = selectedStep?.completed;
-
-    return this.isLinear && !stepCompleted && !formValid;
-  }
+  public constructor(private readonly subscriptionLifecycle: SubscriptionLifecycle) {}
 
-  // TODO: can be made a directive
-  public isSubmitDisabled(stepper: MatStepper): boolean {
-    const selectedStep = stepper.steps.get(stepper.selectedIndex);
-    const formValid = selectedStep?.stepControl?.valid;
-    const stepCompleted = selectedStep?.completed;
+  public ngAfterContentInit(): void {
+    this.steps$ = queryListAndChanges$(this.steps).pipe(map(list => list.toArray()));
 
-    return !stepCompleted && !formValid;
+    /**
+     * Inorder to get the next/submit button disabled status, we need to listen to the primary triggers for change:
+     * 1. Stepper Initialization
+     * 2. Selection Change
+     *
+     * When either of these happens, we need to check if a step control is provided for the step. If a step control is provided, we
+     * switch to the form status changes inorder to find if the form is valid or not.
+     *
+     * If there is no step control provided, we can check for the step completed property.
+     *
+     * Since the same observable is used for both the disabled handling of submit and the next button,
+     * we switch between two logic:
+     * - For next: If the stepper is linear and the step is valid
+     * - For Submit: If all the steps are valid
+     */
+    const sub = merge(
+      this.selectionChange.pipe(map(({ selectedStep }) => selectedStep)),
+      this.steps$.pipe(map(([firstStep]) => firstStep))
+    )
+      .pipe(
+        switchMap(step => {
+          if (step.stepControl) {
+            return step.stepControl.statusChanges.pipe(
+              startWith(step.stepControl.status),
+              map(status => status === 'VALID')
+            );
+          }
+
+          return of(step.completed);
+        }),
+        debounceTime(100) // <-- debounce to avoid multiple emits
```

Comment:
do we need this? if we are only looking for distinct values, we could use distinct or similar operators.

Reply:
Since we are using the `this.step` in the subscribe, in order to get the latest state, we need to fetch the state in the next tick. Hence added a debounce which will prevent multiple emits as well as tick, so we get the current state in subscribe.

Reply:
is there a possibility the debounce will mask the last set of relevant changes within the 100ms interval? Debounce does not replay the new values later, it ignores anything within the 100ms period.

Reply:
![image](https://user-images.githubusercontent.com/29002760/193507902-174e4f88-f87a-4bfc-8b99-0fe45b390ed2.png)

It waits till that mentioned time and emits the last value. So I think we should be fine.

Reply:
IMO looks good but why the 100ms, don't we have a default for this kind of case?

Reply:
I'm not sure. @arjunlalb ?

Reply:
this is a case we haven't encountered before. so, there's no such established defaults. I'm also torn about this approach. I feel like there may be a better way to do this.

Reply:
@aaron-steinfeld any thoughts here?

Reply:
discussed briefly on a chat - summary was basically to prefer using a function over a subject and subscription here to sidestep the issue. If function is called extra times, that's likely not going to be a problem perf wise even if not ideal (and likely able to be memoized away).

As long as there's no human-discernible perf impact, we typically should value readability over the number of clock cycles.

Reply:
@aaron-steinfeld Tried with methods instead. But the problem is that the methods are not called unless there is a change that is detected. The only right trigger here is the `statusChanges` for a form that changes when the form validity changes and we should be using that trigger to re-evaluate the `disabled` status.

Reply:
updating with conversation from slack. Sounds like the issue is the caller is passing in a control object, mutating it and not updating the ref, and we're expecting UI updates.

> so I think there’s a few ways to approach it.
>
> Ideally we change callers to pass in a new object when there’s a change. That’s the whole point of onPush, so we’re kind of misusing it there. Not sure how difficult that is given the form integration.
>
> The other option that I think would still be significantly simpler would just add some explicit change detection - for example, in stepper tab you could maintain a subscription like
> `stepControl?.statusChanges.subscribe(() => this.cdr.markForCheck()`
> would just need to make sure to manage the subscription on the way in (in onChanges) and either subs lifecycle or onDestroy.

Reply:
https://github.com/hypertrace/hypertrace-ui/pull/1835#issuecomment-1270027689